requests-for-comments/2025-03-14-dw-testing-production-data.md

```diff
@@ -0,0 +1,89 @@
+# RFC - Test Pipeline Changes on Real Customer Data w/o Affecting Actual Prod Data
+
+## Background
+We now have reasonable traction with our data warehouse product and the cost of making breaking changes with
+incremental updates is growing as we accrue more and more external sources, self-managed sources, and saved queries each
+day. At this time, we don't have enough confidence in our (mostly unit) test suite and development workloads don't
+closely mimic the realities of production scenarios. Not all the sources have similar distributions of use, but one
+common theme is that many are SQL based and our data eventually end up in s3 via delta tables and are queryable via
+HogQl, so that feels like the area to hone our focus on for biggest impact. Other data sources like HubSpot, Salesforce,
+and Chargebee are less generic and also would have a smaller blast radius in the event syncing incremental or full table
+syncing gets regressed. These sources are still important and should be incorporated into the test plan as it matures.
+
+## Higher Level Goals
+- Determinism - We do not want to randomly sample sources/schemas in order to test against. This list of tests should
+  grow over time, but remain consistent for feature development.
+- Repeatability - A failure one run should fail the same way the next run, A success should also be repeatable.
+- Not Ignorable - A scheduled out of band test can and will be ignored, so this test sequence should run as part of CI.
+- Uses "real" data - Not anchored to "test" datasets that fail to capture the madness of the real world.
+- No Perf Degradation - We don't wish to interfere with any production workloads or queries.
+- No Cost/CPU Impact - We don’t wish to run any out of band testing workloads that would impact a live service, with the
+  possible exception of testing our own services that are in the purview of PostHog.
+
+## Types of Actions to Test
+- Syncs
+    - Incremental
+    - Full table
+- HogQl queries (joins)
+
+> Preface: The main idea behind this strategy is to reverse the direction of our standard syncing procedure in order to
+> determine if we can reconstruct our block storage data (delta → s3) back into the original data store it came from (
+> Postgres, MySql, etc…), and query it once it’s reconstructed as before once it has been put back together.
+
+> Note: The diagram below can be used as an aide.
+
+## Steps
+1. The first step would be to transfer data from a configurable list of sub-buckets under
```

Comment:
How will said "configurable list of sub-buckets" be configured? I assume that we'll need to store the configuration used in each test run to ensure repeatability?